src/main/org/deidentifier/arx/distributed/ARXPartition.java

```diff
@@ -69,164 +100,275 @@ public static Data getData(List<DataHandle> handles) {
-     * @param data
-     * @param number
-     * @return
+     * Splits a given dataset into a specified number of partitions, ensuring that each partition
+     * is within the same equivalence class based on quasi-identifying attributes. Each partition will have
+     * the same header as the original file.
+     *
+     * @param data           The dataset to be partitioned. (Expected to have a header)
+     * @param subsetIndices  A sorted array of indices specifying a subset of the data for partitioning.
+     * @param number         The number of partitions to create.
+     * @return               A list of {@link ARXPartition} objects representing the divided partitions.
      */
-    public static List<DataHandle> getPartitionsByClass(Data data, int number) {
+    public static List<ARXPartition> getPartitionsByClass(Data data, int[] subsetIndices, int number) {
+        DataHandle handle = data.getHandle();
 
-        // Prepare
-        List<DataHandle> result = new ArrayList<>();
         DataDefinition definition = data.getDefinition().clone();
-        DataHandle handle = data.getHandle();
         Set<String> qi = handle.getDefinition().getQuasiIdentifyingAttributes();
         for (String attribute : qi) {
             definition.resetHierarchy(attribute);
         }
-        
-        // Collect indices to use for sorting
-        int[] indices = new int[qi.size()];
-        int num = 0;
-        for (int column = 0; column < handle.getNumColumns(); column++) {
-            if (qi.contains(handle.getAttributeName(column))) {
-                indices[num++] = column;
-            }
-        }
-        
-        // Sort
-        handle.sort(true, indices);
-        
+
+        int[] qiIndices = createIndexArray(handle, qi);
+
+        // Shuffle
+        List<String[]> markedRows = sortData(handle, subsetIndices, qiIndices);
+
+        // Prepare
+        List<ARXPartition> result = new ArrayList<>();
+
         // Split
-        Iterator<String[]> iter = handle.iterator();
+        Iterator<String[]> iter = markedRows.iterator();
         String[] header = iter.next();
-        String[] current = iter.next();
-        String[] next = iter.next();
+        header = Arrays.copyOf(header, header.length - 1); // Remove the marker
+
         int size = (int)Math.floor((double)handle.getNumRows() / (double)number);
-        for (int i = 0; i < number && current != null; i++) {
-            
+        int partitionStartIndex = 0; // To track the start of each partition
+
+        while (iter.hasNext()) {
             // Build this partition
-            List<String[]> _list = new ArrayList<>();
-            _list.add(header);
-            
+            List<String[]> partitionData = new ArrayList<>();
+            List<Integer> partitionSubsetIndices = new ArrayList<>();
+            partitionData.add(header);
+            int partitionEndIndex = partitionStartIndex; // To track the end of each partition
```

Comment:
partitionStartIndex and partitionEndIndex are likely not needed, if sorting is implemented as I suggested above.